File: src/main/java/Q143_ReorderList.java
/***
 * 45mins, debugging silly bug
 * Runtime: 1 ms, faster than 99.94% of Java online submissions for Reorder List.
 * Memory Usage: 44.3 MB, less than 5.04% of Java online submissions for Reorder List.
 */
public class Q143_ReorderList {
    ListNode curr;
    ListNode h;
    int count = 0;
    public void reorderList(ListNode head) {
        curr = new ListNode();
        ListNode tmp = curr;
        h = head;
        rec(head);
        tmp.next = null;
    }

    public void rec(ListNode head) {
        if(head == null) {
            return;
        }
        count++;
        rec(head.next);
        if(count > 1) {
            curr.next = h;
            h = h.next;
            curr = curr.next;
            curr.next = head;
            curr = curr.next;
            curr.next = null;
            count -= 2;
        }
        else if(count == 1) {
            curr.next = head;
            head.next = null;
            count--;
        }
    }
}

File: LeetCodePractice/LeetCodePractice/Q143_ReorderList.cpp
#include "LCHeader.h"
#include "ListNode.cpp"


class Solution {//12:14 12:23
public:
    ListNode* h;
    ListNode* p;
    void reorderList(ListNode* head) {
        ListNode dummy;
        p = &dummy;
        h = head;
        rec(head);
    }

    void rec(ListNode* cur) {
        if (cur == nullptr) {
            return;
        }
        rec(cur->next);
        if (p == nullptr) {
            return;
        }
        if (cur == h) {
            p->next = cur;
            cur->next = nullptr;
            p = nullptr;
        }
        else if (h->next == cur) {
            p->next = h;
            cur->next = nullptr;
            p = nullptr;
        }
        else {
            p->next = h;
            h = h->next;
            p = p->next;
            p->next = cur;
            p = p->next;
            p->next = nullptr;
        }
    }
};